### PR TITLE
Fix upgrade crashing when multiple versions of `fileutils` installed

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## What was the end-user or developer problem that led to this PR?

If you have multiple versions of `fileutils` installed on a ruby where fileutils is a default gem, upgrading rubygems partially fails with an error like this.

```
$ ruby setup.rb             
  Successfully built RubyGem
  Name: bundler
  Version: 2.3.0.dev
  File: bundler-2.3.0.dev.gem
ERROR:  While executing gem ... (NoMethodError)
    undefined method `name' for nil:NilClass
```
 
## What is your fix for the problem, implemented in this PR?

The issue is tricky.  There's 3 relevant moments:

* When first loading the `update` command,  `tmpdir` is required. This require causes the `Gem::Specification.stubs` array to be loaded, because since there are more than one version of `fileutils`, which is a `tmpdir` dependency, `rubygems` needs to figure out which one to require, and needs to load stubs from disk for that.
* Later on, when installing the upgraded default bundler gem, the upgrader first deletes the existing default bundler gemspec from disk before installing the upgraded version.
* Finally, when plugins are being regenerated specification stubs are materialized into actual `Gem::Specification`'s, but since one of them is no longer on disk, it results on a `nil` specification being loaded, causing a runtime error later on when calling a specification method on it.
    
I can think of several ways of fixing this issue:
* Delete the existing spec after installing the new one, not before.
* Require `tmpdir` lazily.
* Don't materialize specs at all when regenerating plugins, since I don't think it's needed. All we need is their versions and names, which stubs already provide.
 
The PR implements the second and third solution. Since I was at it, I removed the specification mutex that I think it was no longer necessary after the third solution. In any case, it was added to support multiple concurrent `gem install` commands, but that has never really been supported, because the gem installer first completely wipes out any existing installation, and then unpackages files from scratch, and that's obviously not thread safe.

As a consequence of this PR, upgrading rubygems should be faster. On my system, it shows a consistent 3% speed up but that's with no gems installed, I guess it could be bigger with many gems installed.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
@simi We got some flaky failure in CI and this PR is very likely to fix it so I'm going to sneak it into the release if you're ok.

**Comment by deivid-rodriguez** [comment]
Actually, I thought this could save us including some regression in this release, but https://github.com/rubygems/rubygems/pull/5050 which I was thinking could've introduced this issue, has already been released. So it's ok to ship this later.

Also, [the flaky failure I'm talking about](https://github.com/rubygems/rubygems/runs/4445320581?check_suite_focus=true( is against an older rubygems, so fixing it would require further patches to bundler.

So to sum up, this PR would be good to merge, but there's no rush to release it.

**Comment by deivid-rodriguez** [comment]
Thanks!!
